File: 242_valid_anagram.js
var isAnagram = function(s, t) {
    if (s.length === 1 && t.length === 1){
        if (s[0] === t[0]){
            return true;
        } else {
            return false;
        }
    }
    if (s.length !== t.length){
        return false;
    }

    const storage = {};

    for (let i = 0; i < s.length; i++){
        let letter = s[i];
        if (storage[letter] !== undefined){
            storage[letter]++;
        } else {
            storage[letter] = 1;
        }
    }

    for (let i = 0; i < t.length; i++){
        let letter = t[i];
        if (storage[letter] !== undefined && storage[letter] > 0){
            storage[letter]--;
        } else {
            return false;
        }
    }
    return true;
    //Time Complexity O(n), Space Complexity O(1)
};

File: 242_valid_anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) == 1 and len(t) == 1:
            if s[0] == t[0]:
                return True
            else:
                return False
        if len(s) != len(t):
            return False
        
        storage = {}

        for i in range(len(s)):
            letter = s[i]
            if letter in storage:
                storage[letter]+=1
            else:
                storage[letter]=1
        
        for i in range(len(t)):
            letter = t[i]
            if letter in storage and storage[letter] > 0:
                storage[letter]-=1
            else:
                return False
        
        return True